JOSE_paper/paper.md

```diff
@@ -0,0 +1,223 @@
+
+
+
+# Teaching materials
+
+All of the teaching materials for this course are available online in a GitHub repository.
+
+
+## Slides
+
+All of the key material is contained in a slide deck for the course available online
+and linked from the repository README.
+The slides are written in [Quarto](https://quarto.org/) markdown [@Allaire_Quarto_2022]
+and rendered as [reveal.js](https://revealjs.com/) html.
+Source and instructions on how to render are included in the repository should
+others wish to tailor them to their specifications.
+
+The slides are broken into separate files by section covering:
+
+- Introduction to research software engineering, why it matters, and some examples,
+- The use of virtual environments and specifying project dependencies,
+- Code formatting for consistent readable layout,
+- Naming principles for readable, re-useable code,
+- Linting and static analysis of code,
+- Writing docstrings and comments, and
+- Other general principles including magic numbers, avoiding hardcoded values, etc.
+
+This makes it easy to remove sections to tailor the course, or for additional topics to be added in future.
+
+## Exercises
+
+The slides are interspersed with a series of exercises, each one applying techniques
+that have just been taught.
+
+We set up a scenario where users have been provided a working, but poorly written piece
+of code that they are now expected to re-use in their work.
+As they progress through the exercises they apply their new skills to transform this
+into a much more readable and understandable piece of code.
+
+The script we use is based on some of the code and data available in @Irving2019.
+This was chosen as it features typical applications we expect users to encounter
+(reading from a dataset, processing, and plotting).
+The course was originally designed to be taught to geoscientists, and then at a climate
+science summer school, and we wanted participants to be focussed on the changes to the
+code rather than worrying about 
+
+Exercises roughly match the slide sections above:
+
+- 01: Baseline code to examine, install dependencies, and run.\
+  Users can familiarise themselves with what the code does, see why listing dependencies
+  with a quick setup is useful, and make early observations about how the code is
+  difficult to understand/use.
+- 02: Apply a formatter to standardise code.\
+  We use the black python formatter [@LangaBlack]
+- 03: Improve code clarity with naming and source changes.
+- 04: Linting and static analysis of code.
+- 05: Writing docstrings and best use of comments.
+- 06: General techniques for better code (magic numbers, string formatting).
+- 00: The end point of the workshop - an improved version of the code in exercise 01.
+
+
+### Solutions
+
+Worked solutions to all of the exercises are provided in the form of completed
+notebooks including example output.
+These are available both in the repository and also as Colab instances.
+
```

Comment:
Add that the exercises build on each other without needing to complete all -- the example code for an exercise will always contain the solution to the previous one

Reply:
Done in https://github.com/jatkinson1000/power-up-python/pull/14/commits/2816d304cbae61641b9a2d4d75ea7fc8b2954b4d